packages/gatsby/src/db/node-tracking.js

```diff
@@ -111,32 +112,25 @@ exports.trackInlineObjectsInRootNode = trackInlineObjectsInRootNode
 const findRootNodeAncestor = (obj, predicate = null) => {
   const { getNode } = require(`./nodes`)
 
-  // Find the root node.
-  let rootNode = obj
-  let whileCount = 0
-  let rootNodeId
-  while (
-    (!predicate || !predicate(rootNode)) &&
-    (rootNodeId = getRootNodeId(rootNode) || rootNode.parent) &&
-    ((rootNode.parent && getNode(rootNode.parent) !== undefined) ||
-      getNode(rootNodeId)) &&
-    whileCount < 101
-  ) {
-    if (rootNodeId) {
-      rootNode = getNode(rootNodeId)
-    } else {
-      rootNode = getNode(rootNode.parent)
-    }
-    whileCount += 1
-    if (whileCount > 100) {
-      console.log(
-        `It looks like you have a node that's set its parent as itself`,
-        rootNode
-      )
-    }
+  let iterations = 0
+  let node = obj
+
+  while (iterations++ < 100) {
+    if (predicate && predicate(node)) return node
+
+    const parent = node.parent && getNode(node.parent)
+    const id = getRootNodeId(node)
+    const trackedParent = id && getNode(id)
+
+    if (!parent && !trackedParent) return node
+
+    node = parent || trackedParent
```

Comment:
if I understand correclyt this is fixing the bug? Rest is just cleanup of the crazy while? 😛

Reply:
yes exactly